src/main/java/com/cloudbees/plugins/credentials/domains/CredentialSet.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2015, Jacob Keller.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.cloudbees.plugins.credentials.domains;
+
+import com.cloudbees.plugins.credentials.Credentials;
+import com.cloudbees.plugins.credentials.CredentialsMatcher;
+import com.cloudbees.plugins.credentials.matchers.AnyOfMatcher;
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.TreeMap;
+import java.util.Map;
+import java.util.List;
+
+/**
+ * Represents a mapping between {@link Domain}s and {@link Credentials},
+ * similar to {@link DomainCredentials}, but providing exact credentials for
+ * each domain.
+ */
+public class CredentialSet extends TreeMap<Domain, Credentials> {
```

Comment:
Ok. I like the idea of dropping the map contract.

Another alternative is to actually extend the BaseCredentials so that a CredentialsSet could actually be treated like a complete set.

What's the difference between a set of DomainSpecifications and a Domain?

I don't really know if we want to extend credentials because I don't think it really honors the domain in the same way.. ie.. you can't really put credentialsSet inside a single domain... but then how does one perform lookups for which set you're going to use?

Generally, I think the use case of a CredentialsSet is something like:

"Given this domain, use this credentials" vs the Domain which is "here is a list of credentials you can choose from if you match this domain".

This is a stronger contract and easier to use. In addition, a CredentialSet enables a collection of domains and their given credentials.

I think the primary connection to them from the use side of credentials should be the "resolveCredentials" as above.

Makes sense to use the string instead of the credentials object itself.

The main reason I implemented it as a contract of a map was to make it easier for the supplying side of credentials to add and modify the mappings.. But maybe that's not really important... We can simply provide a constructor instead that takes a Map.